Remove SNR triggers not close to injection with injfilterrejector

spxiwh opened the pull request:
This PR adds a new option to injfilterrejector such that triggers in the SNR timeseries that are not close to injections are rejected immediately after matchedfiltering and are not sent on the the signal-based vetoes.

This is particularly relevant for pyGRB where the chi-squared can be a dominant cost, and computing this for glitches nowhere near injections in injection runs is unnecessary. For now I only hook this up to multi_inspiral, but it can be easily attached to other inspiral front-ends as needed. Related to #5148 

## Standard information about the request

This is a: new feature, efficiency update, 

This change affects: PyGRB

This change changes: scientific output (although not meaningfully as such triggers would be later removed in post-processing anyway).

This change: follows style guidelines (See e.g. [PEP8](https://peps.python.org/pep-0008/)), has been proposed using the [contribution guidelines](https://github.com/gwastro/pycbc/blob/master/CONTRIBUTING.md)

This change will: not break current functionality, not require additional dependencies, not require a new release

## Motivation

#5148 Lays out the motivation for optimizing pyGRB. This is a part of that.

## Contents

New methods are added to injfilterrejector to do this including command-line options. This is hooked up to multi_inspiral (only) as well.

## Links to any issues or associated PRs

#5148

## Testing performed

I've tested this on the example provided in #5148 (profile will appear there shortly). Other tests welcome (but pyGRB folks need to give me unit tests if they want me to do anything extensive).

- [./] The author of this pull request confirms they will adhere to the [code of conduct](https://github.com/gwastro/pycbc/blob/master/CODE_OF_CONDUCT.md)

spxiwh committed:
Trig based injfilterrejector windowing

titodalcanton commented on bin/pycbc_multi_inspiral:
```diff
@@ -707,7 +723,7 @@ with ctx:
                 # arithmetic with time_delay_idx, which can be positive
                 # or negative
-                idx[ifo] = ind.astype(np.int32)
+                idx[ifo] = ind
```
Please remove the comment here, as it has been moved above.

titodalcanton commented on pycbc/inject/injfilterrejector.py:
```diff
@@ -150,11 +168,17 @@ class InjFilterRejector(object):
         # Determine if InjFilterRejector is to be enabled
-        if injection_file is None or injection_file == 'False' or\
-                (chirp_time_window is None and match_threshold is None):
+        if (
```
The "else" branch of this if statement is unnecessary, please remove.

spxiwh replied:
Done (will be pushed later with other fixes)

titodalcanton reviewed (changes requested):
Two first minor requests, not a careful review yet.

spxiwh committed:
FIx chisq bug that this uncovers

spxiwh committed:
Tito first comments

spxiwh committed:
Another fix to chisq issue

spxiwh commented:
@titodalcanton A poke on this PR ... I think this addition is generally useful, not just in multi_inspiral. With denser injection sets a lot of the original optimizations (e.g. not analysing segments without injections) are no longer relevant, so avoiding unnecessary costs in injection analysis is desirable.

titodalcanton commented on pycbc/inject/injfilterrejector.py:
```diff
@@ -231,6 +259,95 @@ def from_cli_multi_ifos(cls, opt, ifos):
+        ends = inj_times + window
+        intervals = np.c_[starts, ends]
+
+        # Sort intervals by their start points
+        intervals = intervals[intervals[:, 0].argsort()]
```
Can you not simply sort `inj_times` beforehand (maybe even in `get_inj_end_times()`, so you cache them already sorted)?

titodalcanton commented on pycbc/inject/injfilterrejector.py:
```diff
@@ -231,6 +259,95 @@ def from_cli_multi_ifos(cls, opt, ifos):
+        # Sort intervals by their start points
+        intervals = intervals[intervals[:, 0].argsort()]
+
+        merged_intervals = []
+        if intervals.shape[0] > 0:
+            current_merged_interval = intervals[0].copy()
+            for i in range(1, intervals.shape[0]):
+                if intervals[i, 0] <= current_merged_interval[1]:  # Overlap
+                    current_merged_interval[1] = max(
+                        current_merged_interval[1],
+                        intervals[i, 1]
+                    )
+                else:
+                    merged_intervals.append(current_merged_interval)
+                    current_merged_interval = intervals[i].copy()
+            merged_intervals.append(current_merged_interval) # Add the last one
```
Would igwn-segments not work for this?

titodalcanton commented on pycbc/inject/injfilterrejector.py:
```diff
@@ -231,6 +259,95 @@ def from_cli_multi_ifos(cls, opt, ifos):
+            # its corresponding position in 'is_matching' will become True.
+            is_matching |= curr_interval_matches
+
+        # Get the actual indices where 'is_matching' is True
+        # numpy.where returns a tuple, we need the first element which is the array of indices
```
This comment does not make much sense to me.

spxiwh commented on pycbc/inject/injfilterrejector.py:
```diff
@@ -231,6 +259,95 @@ def from_cli_multi_ifos(cls, opt, ifos):
+        return np.array(merged_intervals)
+
+    # Written together with Google Gemini
+    def find_indices_in_injection_intervals(self, trig_times):
```
There doesn't seem to be an optimized way to do index in segmentlist for a list of indices without using a for loop over the indices (and we assume len(indices) is large, whereas len(injection_windows) is not). So I think it would not be worth changing this code.

spxiwh committed:
Tito comments

spxiwh commented:
@titodalcanton I think I've addressed the comments now, does it all make sense? Going to try and run one of your test scripts now to see if it all works okay as well.

titodalcanton reviewed (approved):
No more comments from me, though some good testing is needed.

spxiwh commented:
Having explored this, I think the underlying problem here is that the clustering algorithm (the `findchirp_cluster` algorithm) is not really suitable for this problem. This algorithm does not guarantee that only 1 event is kept in `window`, can remove additional events if run twice and does not give the same output if the run it with all the times set to `- time`. It's really designed to quickly cut down the number of peaks in a continuous timeseries (where adjacent points are strongly correlated), whereas here, because of multiple sky points, a plot of stat vs time would be very spiky.

There are (slower) algorithms in PyCBC that can handle this. Modifying https://github.com/gwastro/pycbc/blob/d791555f6892eeea6fae79857d95e83adb011fb5/pycbc/events/coinc.py#L483 to return `indices` seems most appropriate. I implemented that for testing and now see:
```
2025-10-03T00:52:16.319-07:00 INFO : Preclustered 260275
2025-10-03T00:52:16.692-07:00 INFO : Postclustered 16
2025-10-03T00:52:16.692-07:00 INFO : Postclustered2 16
2025-10-03T00:52:16.692-07:00 INFO : Postclustered3 16
```
(I ran clustering 3 times to check this number doesn't change). The cost of this doesn't *seem* to be a problem compared to everything else, but I'll do a larger profile if putting this in a separate patch. I guess the main point here is *there's over 2 orders of magnitude fewer triggers post-clustering with a "proper" algorithm*. If seeing issues with things like `finalise_templates` having 2 orders of magnitude fewer triggers will fix this. I also note that coh_PTF did not use `findchirp_cluster` for this (https://lscsoft.docs.ligo.org/lalsuite/lalapps/coh___p_t_f__utils_8c_source.html#l03153) ... I believe this is the "naive" O(N^2) implementation of the coinc.py cluster algorithm in PyCBC.

Anyway, changing the clustering algorithm is out of scope of this patch, but it's hard to compare without. I'll therefore run Tito's test cases, with the new clustering algorithm, and with and without this trig window applied to compare outputs. I hope this is then enough to approve and merge this PR.

pannarale reviewed (approved):
Looks, I was only able to find an extra 's' in a doctoring

spxiwh committed:
Fix gramma-o
